src/java.base/share/classes/jdk/internal/misc/Signal.java

```diff
@@ -173,21 +173,20 @@ public static synchronized Signal.Handler handle(Signal sig,
                 ("Signal already used by VM or OS: " + sig);
         }
         signals.put(sig.number, sig);
-        synchronized (handlers) {
-            Signal.Handler oldHandler = handlers.get(sig);
-            handlers.remove(sig);
-            if (newH == 2) {
-                handlers.put(sig, handler);
-            }
-            if (oldH == 0) {
-                return Signal.Handler.SIG_DFL;
-            } else if (oldH == 1) {
-                return Signal.Handler.SIG_IGN;
-            } else if (oldH == 2) {
-                return oldHandler;
-            } else {
-                return new NativeHandler(oldH);
-            }
+        Signal.Handler oldHandler;
+        if (newH == 2) {
+            oldHandler = handlers.put(sig, handler);
+        } else {
+            oldHandler = handlers.remove(sig);
+        }
```

Comment:
A ternary assignment might be an alternative here:
```
 Signal.Handler oldHandler = (newH == 2) ? handlers.put(sig, handler) : handlers.remove(sig);
```

Reply:
Too much logic in the single line. I like `if` approach more.